### ISSUE TITLE
Exception on parsing a one page PDF

### ISSUE THREAD

**Issue by jksemple** [UglyToad/PdfPig#118]
Hi Eliot,
I've installed PdfPig 0.0.11 from Nuget.

I'm getting an ArgumentOutOfRangeException when calling PdfDocument.GetPages() on a small one page PDF.

The stack trace is:

```
at System.SZArrayHelper.get_Item[T](Int32 index)
   at UglyToad.PdfPig.Fonts.CompactFontFormat.CharStrings.Type2CharStringParser.GetCommand(Byte b, List`1 bytes, List`1 precedingCommands, CompactFontFormatIndex localSubroutines, CompactFontFormatIndex globalSubroutines, Int32& i) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Fonts\CompactFontFormat\CharStrings\Type2CharStringParser.cs:line 803
   at UglyToad.PdfPig.Fonts.CompactFontFormat.CharStrings.Type2CharStringParser.ParseSingle(List`1 bytes, CompactFontFormatIndex localSubroutines, CompactFontFormatIndex globalSubroutines) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Fonts\CompactFontFormat\CharStrings\Type2CharStringParser.cs:line 726
   at UglyToad.PdfPig.Fonts.CompactFontFormat.CharStrings.Type2CharStringParser.Parse(IReadOnlyList`1 charStringBytes, CompactFontFormatIndex localSubroutines, CompactFontFormatIndex globalSubroutines, ICompactFontFormatCharset charset) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Fonts\CompactFontFormat\CharStrings\Type2CharStringParser.cs:line 707
   at UglyToad.PdfPig.Fonts.CompactFontFormat.CompactFontFormatIndividualFontParser.Parse(CompactFontFormatData data, String name, IReadOnlyList`1 topDictionaryIndex, IReadOnlyList`1 stringIndex, CompactFontFormatIndex globalSubroutineIndex) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Fonts\CompactFontFormat\CompactFontFormatIndividualFontParser.cs:line 105
   at UglyToad.PdfPig.Fonts.CompactFontFormat.CompactFontFormatParser.Parse(CompactFontFormatData data) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Fonts\CompactFontFormat\CompactFontFormatParser.cs:line 56
   at UglyToad.PdfPig.Fonts.Parser.Parts.CidFontFactory.ReadDescriptorFile(FontDescriptor descriptor) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Fonts\Parser\Parts\CidFontFactory.cs:line 150
   at UglyToad.PdfPig.Fonts.Parser.Parts.CidFontFactory.Generate(DictionaryToken dictionary, Boolean isLenientParsing) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Fonts\Parser\Parts\CidFontFactory.cs:line 56
   at UglyToad.PdfPig.Fonts.Parser.Handlers.Type0FontHandler.Generate(DictionaryToken dictionary, Boolean isLenientParsing) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Fonts\Parser\Handlers\Type0FontHandler.cs:line 34
   at UglyToad.PdfPig.Fonts.FontFactory.Get(DictionaryToken dictionary, Boolean isLenientParsing) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Fonts\FontFactory.cs:line 54
   at UglyToad.PdfPig.Content.ResourceStore.LoadFontDictionary(DictionaryToken fontDictionary, Boolean isLenientParsing) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Content\ResourceStore.cs:line 137
   at UglyToad.PdfPig.Content.ResourceStore.LoadResourceDictionary(DictionaryToken resourceDictionary, Boolean isLenientParsing) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Content\ResourceStore.cs:line 39
   at UglyToad.PdfPig.Parser.PageFactory.Create(Int32 number, DictionaryToken dictionary, PageTreeMembers pageTreeMembers, Boolean isLenientParsing) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Parser\PageFactory.cs:line 68
   at UglyToad.PdfPig.Content.Pages.GetPage(Int32 pageNumber) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\Content\Pages.cs:line 69
   at UglyToad.PdfPig.PdfDocument.GetPage(Int32 pageNumber) in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\PdfDocument.cs:line 154
   at UglyToad.PdfPig.PdfDocument.GetPages()+MoveNext() in C:\git\csharp\UglyToad.PdfPig\src\UglyToad.PdfPig\PdfDocument.cs:line 174
   at PDFParser.PdfParser.ExtractTextFromFileContents(Byte[] fileContents, String regexPattern) in e:\My Documents\Visual Studio 2017\Projects\Assets\PDFParser\PdfParser.cs:line 33
   at PdfParser.Tests.UnitTest1.ParseContract() in e:\My Documents\Visual Studio 2017\Projects\Assets\PdfParser.Tests\UnitTest1.cs:line 23
Result Message:	
Test method PdfParser.Tests.UnitTest1.ParseContract threw exception: 
System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection.
Parameter name: index
```

**Comment by EliotJones** [comment]
Hi, sorry to hear you're encountering this issue.

Are you able to share the file that causes the issue at all please? It looks like there's an edge case in some of the Type 2 Charstring parsing logic I've not seen before and I'm not sure what it is without the file for reference.

Thanks

**Comment by jksemple** [comment]
Yes. Here you are… I’ve redacted anything sensitive.

Regards

Jon

**Comment by EliotJones** [comment]
Sorry, it looks like the attachment didn't come through on the email, it might be necessary to upload it directly to the GitHub issue, or you can email it directly to me elioty (at) hotmail (dot) co (dot) uk

**Comment by jksemple** [comment]
I sent it to your hotmail address. Did you get it ok?
regards
Jon

**Comment by EliotJones** [comment]
Yes, thank you, I'll take a look today.

**Comment by EliotJones** [comment]
It looks like this was one of the issues fixed by changes on master since 0.0.11, I'm not sure which one. The next release (0.1.0) is going to be within the next few weeks, but will feature some large changes so I'm going to release a couple of pre-release versions first.

In the meantime I created this package for the changes required by #98 but it also includes a fix for the issue you are encountering: [PdfPig.0.0.11.98.zip](https://github.com/UglyToad/PdfPig/files/4021781/PdfPig.0.0.11.98.zip)

**Comment by jksemple** [comment]
That's done the trick! Well done and thanks for such a timely response.
regards
Jon
